Some enhancements

gkorepanov opened the pull request:
1. Extended `.gitignore` to ignore everything but needed files.
Now all the executables, object files and other auxiliaries are ignored.
2. Unpacked the tests as soon as `git` has its own powerful compression mechanisms and its considered outdated to keep archives in repo, since it's quite inconvenient.
3. Added a `mips-binutils` Makefile, which accomplishes [MIPS-binutils installation guide](https://github.com/MIPT-ILab/mipt-mips/wiki/MIPS-binutils) automatically and should run on both MacOS and Linux systems.

gkorepanov committed:
Some optimiztions done

Revised .gitignore, removed archives, added makefile for binutils

gkorepanov committed:
Updated gitignore to unignore makefiles

gkorepanov committed:
Unignored samples

gkorepanov committed:
Written Makefile for binutils

gkorepanov committed:
Fixed some minor problems

gkorepanov committed:
Final modifications for binutils

gkorepanov committed:
Added libelf installer

gkorepanov committed:
Polished gitignores

gkorepanov committed:
Added all phony targets

gkorepanov committed:
Added cross platfrom install and unistall commands for libelf

gkorepanov committed:
Added sudo's

gkorepanov committed:
Removed deleting the .s files

pavelkryukov commented on .gitignore:
```diff
@@ -1 +1,51 @@
+# git has its own built in compression methods
+*.7z
+*.dmg
+*.gz
```
Wrong. We keep traces in .gz

pavelkryukov commented on tests/samples/factorial.s:
```diff
@@ -0,0 +1,112 @@
+#return 1;
+returnOne:
+  li      $t0, 1            # load 1 into register $t0
+  sw      $t0, 4($sp)       # store 1 into the parent's return value register
```
I have INTENTIONALLY compressed all samples file so they won't interfere with actual sources.
It is a common practice to keep traces compressed as we would have many traces on our disks and unzip only few of them.

gkorepanov replied:
Got it. If needed, I might revert this, but seems that won't make sense as we don't need those changes at all.

gkorepanov replied:
It was just my fault not to track the history of those gzipped traces. Sorry for that.

pavelkryukov commented:
I don't see much value in these `polishes`. "Premature optimization is the root of all evil".
We have a lot of other things to do which have more priority.

gkorepanov committed:
Revised README in tests dir

pavelkryukov commented on libs/mips-binutils/Makefile:
```diff
@@ -1,22 +1,66 @@
 
+DIR= binutils
+ARCHIVE= binutils-2.28.tar.bz2
+URL= http://ftp.gnu.org/gnu/$(DIR)/$(ARCHIVE)
```
I don't like the idea of automatized installation (both libelf and binutils) as nobody will maintain that code after you leave the project. However, manual installation is a good thing as an exercise for students, and in case of problem they may ask mentor and/or fix the Wiki page.

gkorepanov replied:
That's clear. Isn't it true for google test as well?

pavelkryukov replied:
I don't like the way GoogleTest is included to the repository now (through wget). It is OK as we do not have anything better, but it is not good to copy that for other libraries.

The good thing about GoogleTest that it is not installed deeply inside your machine. Assume I am a external user who wants to try MIPT-MIPS. I played with it a little bit, I didn't like it and removed it completely by `rm -rf`. However, this installers did some implicit changes in Linux which may cause some unexpected consequences. For example, are you sure that installer would work correctly if previous version of libelf is installed? If _newer_ version is installed?

Another feature of GoogleTest installation is a real cross-platformness. I am not sure that your Makefiles would work on Solaris, BSD or something exotic, while manual can be easily adjusted by reader himself/herself.

gkorepanov replied:
Now I understand the difference. Thank you very much for clear explanation.

pavelkryukov commented on libs/mips-binutils/Makefile:
```diff
@@ -68,7 +68,7 @@ clean:
 
 uninstall_mips-binutils:
-	rm -rf $(ALIASES) $(PREFIX)
+	sudo rm -rf $(ALIASES) $(PREFIX)
```
Surprise!

gkorepanov commented:
They are not even polishes, but some obvious things. Just added the Makefiles to complete the build system and added gitignores since it is quite inconvenient to work when there are so many files tracked by `git`

gkorepanov commented:
I don't mean these changes to be taken in account somehow, just added them for purposes of convenience hoping that some of them would be accepted. So if you consider them completely useless, please, just close the PR.

pavelkryukov commented:
Is there any bug opened with concerns about gitignores? No, because it is not the first priority issue now. We have so many problems with simulation, so please spend your resources to deal with them and my resources to review the changes.
